### PR TITLE
Refactorization

### PR DESCRIPTION
**Pull request by oliverpool**
Hi,

this implements most of #1.

I didn't have time to dig into your queue implementation : do you think you could write it as a [`queue.Interface`](https://github.com/chmike/EChan/compare/master...oliverpool#diff-64388b856a1f68d7e50f7a937f450f4dR8) ?
```go
type Interface interface {
	Push(interface{}) bool    // return false if queue is full (item was not pushed)
	Pop() (interface{}, bool) // return false if queue is empty
}
```

We could also add some standard tests for queue (capped or not) and maybe some benchmarks as well!

### CODE REVIEW

**Comment by chmike** [comment]
Yes, but I can't work on it now. I have job work to do. Sorry.

**Comment by chmike** [comment]
I have some time again to work en EChan.

The rational to create EChan was to have an elastic channel. Instead of allocating a big fixed size buffered channel, the buffer would grow and shrink as needed. The use case is when a program has many such channels and they may fill up for a short time on rare occasion when a transient blocking occurs in one of the consumers, and blocking the producer is not an option (events may get lost). A maximum capacity limit is still needed to avoid that the program or the computer explodes. 
The go routine is needed to grow and shrink the intermediate buffer as needed.

Do you see other use cases for the code you wrote that I missed ?

**Comment by chmike** [comment]
OK. I understand what you did. It is impressive. That is Go level 100. I'm only level 10.  

> How do you determine this limit ?
I would determine it based on the average data production rate and the acceptable duration of a consumer transient block.

I'll call it ering for elastic ring. May I add it to your branch or should I create another branch ?

**Comment by oliverpool** [comment]
> That is Go level 100. I'm only level 10.

:smile: 

> May I add it to your branch or should I create another branch ?

As you prefer. Maybe you can make another branch based on mine and open a pull request to merge it on this branch : this way it is easier to review it and comment it independently

**Comment by oliverpool** [comment]
PS:  I based my structure on this great blog post: https://medium.com/@benbjohnson/standard-package-layout-7cdbc8391fc1

**Comment by oliverpool** [comment]
Before merging this PR, we should remove most of the code from `echan.go` and update the Readme.

The next step (in another PR) would be to do some benchmarking to compare the different solutions !

File: queue/ring/ring.go [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+	r.head = (r.head + 1) % r.cap
+	r.len++
+	return true
+}
+
+func (r *ring) Pop() (x interface{}, ok bool) {
+	if r.len == 0 {
+		return nil, false
+	}
+	x = r.buf[(r.head-r.len+r.cap)%r.cap]
```
Comment (oliverpool) [inline queue/ring/ring.go]: I should set `r.buf[(r.head-r.len+r.cap)%r.cap]` to nil, for it to be garbage collected (as you did in your ering) !

**Comment by chmike** [comment]
OK with me. 

May be we should create a new project with a  new name.  First it's not good to have upper case letter in the name. Second, the repository name should reflect the purpose or main functionality of the tool. 

I was initially only looking for an elastic channel. What you conceived can do more than that. It can bind two existing channels together.

**Comment by oliverpool** [comment]
I agree with you regarding the name:
I like the sentence "Package ... provides a way to insert a buffer between two channels."
Maybe `bchan` is a better name ?

I added some benchmarks and it seems that the perf differences aren't very significant... :smile:

**Pull request merged**